### PR TITLE
[pipeline] Add Pipeline Forward for GPT2Model Shardformer

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4200

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Support pipeline forward for transformers.GPT2Model and add relevant tests.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/policies/gpt2.py [hunk 1]
```diff
@@ -194,3 +242,222 @@ class GPT2ForSequenceClassificationPolicy(GPT2Policy):
 
     def __init__(self) -> None:
         super().__init__()
+
+
+class GPT2PipelineForwards:
```
Comment (CjhHa1) [inline colossalai/shardformer/policies/gpt2.py:247]: so, use class and '' or class and import here ?

File: colossalai/shardformer/policies/gpt2.py [hunk 2]
```diff
@@ -194,3 +242,222 @@ class GPT2ForSequenceClassificationPolicy(GPT2Policy):
...
+                input_shape = inputs_embeds.size()[:-1]
+                batch_size = inputs_embeds.shape[0]
+            else:
+                raise ValueError("You have to specify either input_ids or inputs_embeds")
+
+            device = input_ids.device if input_ids is not None else inputs_embeds.device
+            if token_type_ids is not None:
+                token_type_ids = token_type_ids.view(-1, seq_length)
+        else:
+            assert hidden_states is not None
```
Comment (CjhHa1) [inline colossalai/shardformer/policies/gpt2.py:302]: use if and raise error is better

**Review by CjhHa1 — COMMENTED** [review]
LGTM

File: colossalai/shardformer/policies/gpt2.py (hunk 1 above)
Comment (Fridge003) [reply colossalai/shardformer/policies/gpt2.py:247]: I will use class and '' for typehint.

File: colossalai/shardformer/policies/gpt2.py (hunk 2 above)
Comment (Fridge003) [reply colossalai/shardformer/policies/gpt2.py:302]: Modified.

**Review by CjhHa1 — APPROVED** [review]

**Pull request merged**